hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndexV2.java

```diff
@@ -0,0 +1,223 @@
+import scala.Tuple2;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Simplified re-implementation of {@link HoodieBloomIndex} that does not rely on caching, or
+ * incurs the overhead of auto-tuning parallelism.
+ */
+public class HoodieGlobalBloomIndexV2<T extends HoodieRecordPayload> extends HoodieBloomIndexV2<T> {
+
+  private static final Logger LOG = LogManager.getLogger(HoodieGlobalBloomIndexV2.class);
+
+  public HoodieGlobalBloomIndexV2(HoodieWriteConfig config) {
+    super(config);
+  }
+
+  @Override
+  public JavaRDD<HoodieRecord<T>> tagLocation(JavaRDD<HoodieRecord<T>> recordRDD,
+                                              JavaSparkContext jsc,
+                                              HoodieTable<T> hoodieTable) {
+    return recordRDD
+            .sortBy((record) -> String.format("%s-%s", record.getPartitionPath(), record.getRecordKey()),
+                    true, config.getBloomIndexV2Parallelism())
+            .mapPartitions((itr) -> new LazyRangeAndBloomChecker(itr, hoodieTable)).flatMap(List::iterator)
+            .sortBy(Pair::getRight, true, config.getBloomIndexV2Parallelism())
+            .mapPartitions((itr) -> new LazyKeyChecker(itr, hoodieTable))
+            .filter(Option::isPresent)
+            .map(Option::get);
+  }
+
+  /**
+   * This is not global, since we depend on the partitionPath to do the lookup.
+   */
+  @Override
+  public boolean isGlobal() {
+    return true;
+  }
+
+  /**
+   * Given an iterator of hoodie records, returns a pair of candidate HoodieRecord, FileID pairs,
+   * by filtering for ranges and bloom for all records with all fileIds.
+   */
+  class LazyRangeAndBloomChecker extends
+          LazyIterableIterator<HoodieRecord<T>, List<Pair<HoodieRecord<T>, String>>> {
+
+    private HoodieTable<T> table;
+    private List<Pair<String, String>> partitionPathFileIDList;
+    private IndexFileFilter indexFileFilter;
+    private ExternalSpillableMap<String, BloomFilter> fileIDToBloomFilter;
+    private HoodieTimer hoodieTimer;
+    private long totalTimeMs;
+    private long totalCount;
+    private long totalMetadataReadTimeMs;
+    private long totalRangeCheckTimeMs;
+    private long totalBloomCheckTimeMs;
+    private long totalMatches;
+
+    public LazyRangeAndBloomChecker(Iterator<HoodieRecord<T>> in, final HoodieTable<T> table) {
+      super(in);
+      this.table = table;
+    }
+
+    @Override
+    protected List<Pair<HoodieRecord<T>, String>> computeNext() {
+
+      List<Pair<HoodieRecord<T>, String>> candidates = new ArrayList<>();
+      if (!inputItr.hasNext()) {
+        return candidates;
+      }
+
+      HoodieRecord<T> record = inputItr.next();
+
+      // <Partition path, file name>
+      hoodieTimer.startTimer();
+      Set<Pair<String, String>> matchingFiles = indexFileFilter
```

Comment:
correct me if I am wrong. in global index, there are chances that incoming record key is updated to a different partition path compared to existing record. For example, existing record in hoodie could be <rec_key1, partitionPath1>, and incoming record could be <rec_key1, partitionPath2>. We have to find such locations as well in global. 
if we call indexFileFilter only for the partitionpath and recordkey from incoming record, then we might miss this case.

Reply:
Guess, you missed the config.getBloomIndexUpdatePartitionPath(). Check what HoodieGlobalBloomIndex does for this config. if need be, I can explain too.

Reply:
bottom line, we need to consider all partitions, not just the ones where incoming records are tagged with.